pkg/api/v1/control-plane/cache/prioritylist.go

```diff
@@ -0,0 +1,165 @@
+const defaultPrioirty = -1
+
+type PriorityValue interface {
+	GetPriority() string
+}
+
+// PrioritySortedStruct sorts elements by prioirity. Priority lists are unordered.
+// Needs a priority list and a get priority function.
+// If a priority does not exist for an added element, it is added to the lowest priority.
+type PrioritySortedStruct struct {
+	// prioritySets is the set lists for priorities, where P is the type used for priority.
+	prioritySets map[int][]string
+	// priorityMap maps the value to the priority.
+	priorityMap map[string]int
+	// elements are the map of elements structured by their priority.
+	elements []map[int]ResponseWatch
+	// numberOfPriorities is the number of priorities in the prioritySets
+	numberOfPriorities int
+	// nextUniqueElementIndex is the next element index to be inserted.
+	nextUniqueElementIndex int
+}
+
+// PriorityIndex is the priority and index used to locate items.
+type PriorityIndex struct {
+	Priority int
+	Index    int
+}
+
+func NewPrioritySortedStruct(prioritySets map[int][]string) *PrioritySortedStruct {
+	p := PrioritySortedStruct{
+		prioritySets: prioritySets,
+	}
+	p.Init()
+	return &p
+}
+
+func (p *PrioritySortedStruct) Init() {
+	p.numberOfPriorities = len(p.prioritySets)
+	// need to ensure that the prioriries are in order and there are no missing or skipped Priorities
+	priorities := make([]int, 0)
+	for priority, _ := range p.prioritySets {
+		priorities = append(priorities, priority)
+	}
+	sort.Ints(priorities)
+	currentP := 0
+	for _, p := range priorities {
+		if currentP == p {
+			currentP++
+		} else {
+			panic(fmt.Sprintf("Priorities are not set correct, you are missing priority %d", currentP))
+		}
+	}
+	p.elements = make([]map[int]ResponseWatch, p.numberOfPriorities+1)
+	if p.prioritySets != nil && p.numberOfPriorities > 0 {
+		for priorityIndex := range p.prioritySets {
+			p.elements[priorityIndex] = make(map[int]ResponseWatch)
+		}
+		p.elements[p.numberOfPriorities] = make(map[int]ResponseWatch)
+	} else {
+		// there is only one map of elements
+		p.elements[0] = make(map[int]ResponseWatch)
+	}
+	p.priorityMap = make(map[string]int)
+	for index, pl := range p.prioritySets {
+		for _, v := range pl {
+			p.priorityMap[v] = index
+		}
+	}
+	p.nextUniqueElementIndex = 0
+}
+
+// Get returns the element at the index, and if it exists.
+func (p *PrioritySortedStruct) Get(pi PriorityIndex) (ResponseWatch, bool) {
+	v, ok := p.elements[pi.Priority][pi.Index]
+	return v, ok
+}
+
+// Process will call the procesFunc over all the elements by priority.
+func (p *PrioritySortedStruct) Process(processFunc func(el ResponseWatch, pi PriorityIndex)) {
+	for i := 0; i <= p.numberOfPriorities; i++ {
+		m := p.elements[i]
+		for index, v := range m {
+			processFunc(v, PriorityIndex{Priority: i, Index: index})
+		}
+	}
+}
+
+// GetPriorityList returns an ordered list of the elements by priority.
+func (p *PrioritySortedStruct) GetPriorityList() []ResponseWatch {
```

Comment:
If we are recreating the list each time anyways perhaps we go with the simple section where we just recreate the list from the existing map? It seems like we are adding overhead and complexity here while also diverging more from upstream. 
This function makes me question if I understand what we are trying to do.

Reply:
This is not used at all in the code.
This is if the user of the code wanted to `GetPriorityList` from the struct

Look at the [PR](https://github.com/solo-io/go-utils/pull/483)